### PR TITLE
fix GEN-524 by equilibrating argdiff shape in `VmapTrace`

### PR DESCRIPTION
**Pull request by littleredcomputer**
- correct an imbalance of argdiffs vs primals in VmapTrace
- stop auto-wrapping Flag around MaskTrace flag argument
- add Rif's failing example as unit test

### CODE REVIEW

**Comment by littleredcomputer** [comment]
@femtomc Please look at the one line change in `vmap.py`. The `Mask.update` does a `cond` between `update_change_target` and `update_change_target_from_false`. One of these is routed to `VmapCombinator.update_importance` which puts the argdiffs in the trace, and the other is routed to `VmapCombinator.update_choice_map`, which puts the **primals** in the trace.  `Cond` requires that this choice be made consistently. Perhaps the "opposite" change is better?

**Comment by femtomc** [comment]
The argdiffs should not go in the trace -- whichever option follows that pattern is the correct one.

File: src/genjax/_src/generative_functions/combinators/mask.py [hunk 1]
```diff
@@ -121,8 +121,11 @@ def simulate(
         args: tuple[Any, ...],
     ) -> MaskTrace[R]:
         check, inner_args = args[0], args[1:]
+        if not isinstance(check, Flag):
+            check = Flag(check)
```
Comment (sritchie) [inline src/genjax/_src/generative_functions/combinators/mask.py]: Is there some constructor function we could write that would automatically handle this flattening? We have something like it in `Mask.maybe`...

File: src/genjax/_src/generative_functions/combinators/vmap.py [hunk 2]
```diff
@@ -231,7 +231,7 @@ def _update(key, idx, subtrace, argdiffs):
         w = jnp.sum(w)
         retval = new_subtraces.get_retval()
         scores = new_subtraces.get_score()
-        map_tr = VmapTrace(self, new_subtraces, primals, retval, jnp.sum(scores))
+        map_tr = VmapTrace(self, new_subtraces, argdiffs, retval, jnp.sum(scores))
```
Comment (sritchie) [inline src/genjax/_src/generative_functions/combinators/vmap.py]: as @femtomc notes, `argdiffs` should not go in the trace, so we do in fact want `primals` here.

Comment (sritchie) [reply src/genjax/_src/generative_functions/combinators/vmap.py]: (and some other argdiffs in the file probably needs to change to primals?)

**Review by sritchie — APPROVED** [review]
Nice, approve post-changes

**Comment by sritchie** [comment]
@littleredcomputer can you also look at https://github.com/probcomp/genjax/pull/1261? I feel like @arijitnoobstar 's PR is covering a similar / adjacent problem, it'd be great to get these both in.

**Comment by littleredcomputer** [comment]
> @littleredcomputer can you also look at #1261? I feel like @arijitnoobstar 's PR is covering a similar / adjacent problem, it'd be great to get these both in.

Agree, merging that PR would have saved us some time. Rather than try to rebase it I propose to add @arijitnoobstar as an author of this PR after incorporating the rest of his fix

File: src/genjax/_src/generative_functions/combinators/vmap.py (hunk 2 above)
Comment (littleredcomputer) [reply src/genjax/_src/generative_functions/combinators/vmap.py]: ...perhaps, but we'll want test cases etc.

File: tests/generative_functions/test_mask_combinator.py [hunk 3]
```diff
@@ -116,3 +115,44 @@ def test_mask_update_weight_to_argdiffs_from_false(self, key):
         w = tr.update(key, argdiffs)[1]
         assert w == 0.0
         assert w == tr.get_score()
+
+    def test_mask_scan_update(self, key):
+        def masked_scan_combinator(step, **scan_kwargs):
+            def scan_step_pre(state, flag):
+                return flag, state
+
```
Comment (femtomc) [inline tests/generative_functions/test_mask_combinator.py:123]: worried about this -- but not in this PR (more generally, in `masked_scan`)

It would be easy to generate generative functions where the masked return value could be masked False -- which means unwrapping it and returning the value is invalid (invalid meaning, if its use affects generative computation later on, the scores and weights computed may be incorrect)

**Comment by sritchie** [comment]
@littleredcomputer note that we want to make sure we solve the issue in https://probcomp.slack.com/archives/C06EV7GHCCW/p1724876582854649?thread_ts=1724873178.418879&cid=C06EV7GHCCW.
